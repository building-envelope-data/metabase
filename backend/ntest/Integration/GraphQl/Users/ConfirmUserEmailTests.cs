using System.Threading.Tasks;
using FluentAssertions;
using Snapshooter.NUnit;
using NUnit.Framework;

namespace Metabase.NTests.Integration.GraphQl.Users
{
    [TestFixture]
    public sealed class ConfirmUserEmailNTests
      : UserIntegrationNTests
    {
        [Test]
        public async Task ValidDataWithConfirmationCodeFromRegistrationEmail_ConfirmsUserEmail()
        {
            // Arrange
            var email = "john.doe@ise.fraunhofer.de";
            await RegisterUser(email: email).ConfigureAwait(false);
            var confirmationCode = ExtractConfirmationCodeFromEmail();
            // Act
            var response = await ConfirmUserEmail(
                email: email,
                confirmationCode: confirmationCode
                ).ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                matchOptions => matchOptions.Assert(fieldOptions =>
                 fieldOptions.Field<string>("data.confirmUserEmail.user.id").Should().NotBeNullOrWhiteSpace()
                 )
                );
        }

        [Test]
        public async Task ValidDataWithConfirmationCodeFromResendUserEmailConfirmation_ConfirmsUserEmail()
        {
            // Arrange
            var email = "john.doe@ise.fraunhofer.de";
            await RegisterUser(email: email).ConfigureAwait(false);
            EmailSender.Clear();
            await ResendUserEmailConfirmation(
                email
                ).ConfigureAwait(false);
            var confirmationCode = ExtractConfirmationCodeFromEmail();
            // Act
            var response = await ConfirmUserEmail(
                email: email,
                confirmationCode: confirmationCode
                ).ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                matchOptions => matchOptions.Assert(fieldOptions =>
                 fieldOptions.Field<string>("data.confirmUserEmail.user.id").Should().NotBeNullOrWhiteSpace()
                 )
                );
        }

        [Test]
        public async Task ValidDataWithConfirmationCodeFromResendUserEmailVerification_ConfirmsUserEmail()
        {
            // Arrange
            var email = "john.doe@ise.fraunhofer.de";
            await RegisterAndConfirmAndLoginUser(email: email).ConfigureAwait(false);
            EmailSender.Clear();
            await ResendUserEmailVerification().ConfigureAwait(false);
            var confirmationCode = ExtractConfirmationCodeFromEmail();
            // Act
            var response = await ConfirmUserEmail(
                email: email,
                confirmationCode: confirmationCode
                ).ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                matchOptions => matchOptions.Assert(fieldOptions =>
                 fieldOptions.Field<string>("data.confirmUserEmail.user.id").Should().NotBeNullOrWhiteSpace()
                 )
                );
        }

        [Test]
        public async Task ResentUserEmailConfirmation_ContainsValidCode()
        {
            // Arrange
            var email = "john.doe@ise.fraunhofer.de";
            await RegisterUser(email: email).ConfigureAwait(false);
            EmailSender.Clear();
            // Act
            await ResendUserEmailConfirmation(
                email
                ).ConfigureAwait(false);
            var response = await ConfirmUserEmail(
                email: email,
                confirmationCode: ExtractConfirmationCodeFromEmail()
            ).ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                matchOptions => matchOptions.Assert(fieldOptions =>
                 fieldOptions.Field<string>("data.confirmUserEmail.user.id").Should().NotBeNullOrWhiteSpace()
                 )
            );
        }

        [Test]
        public async Task UnknownUser_IsUserError()
        {
            // Arrange
            var email = "john.doe@ise.fraunhofer.de";
            await RegisterUser(email: email).ConfigureAwait(false);
            var confirmationCode = ExtractConfirmationCodeFromEmail();
            // Act
            var response = await ConfirmUserEmail(
                email: "unknown." + email,
                confirmationCode: confirmationCode
                ).ConfigureAwait(false);
            // Assert
            Snapshot.Match(response);
        }

        [Test]
        public async Task InvalidConfirmationCode_IsUserError()
        {
            // Arrange
            var email = "john.doe@ise.fraunhofer.de";
            await RegisterUser(email: email).ConfigureAwait(false);
            var confirmationCode = ExtractConfirmationCodeFromEmail();
            // Act
            var response = await ConfirmUserEmail(
                email: email,
                confirmationCode: "invalid" + confirmationCode
                ).ConfigureAwait(false);
            // Assert
            Snapshot.Match(response);
        }
    }
}