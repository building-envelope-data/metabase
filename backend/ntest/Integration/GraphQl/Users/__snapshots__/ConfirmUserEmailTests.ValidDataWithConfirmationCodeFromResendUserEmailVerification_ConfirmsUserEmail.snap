{"data":{"confirmUserEmail":{"user":{"id":"VXNlcgpnMjkyM2UyY2M3MzE2NDFiNWI1NTkxYzQ3NDcwYzA1MWM=","email":"john.doe@ise.fraunhofer.de"},"errors":null}}}
